// @react-mdc/drawer/lib/Temporary/Container.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var classNames = require("classnames");
var PropTypes = require("prop-types");
var mdc_drawer_1 = require("@material/drawer/dist/mdc.drawer");
var immutable_1 = require("immutable");
var base_1 = require("@react-mdc/base");
var util_1 = require("@react-mdc/base/lib/util");
var adapter_1 = require("./adapter");
var constants_1 = require("./constants");
var drawerUtil = require("./drawerUtil");
exports.CLASS_NAME = constants_1.BASE_CLASS_NAME;
var OPEN_CLASS_NAME = mdc_drawer_1.MDCTemporaryDrawerFoundation.cssClasses.OPEN, OPACITY_VAR_NAME = mdc_drawer_1.MDCTemporaryDrawerFoundation.strings.OPACITY_VAR_NAME;
var Meta = (function (_super) {
    __extends(Meta, _super);
    function Meta(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            foundationClasses: immutable_1.OrderedSet(),
            foundationCssVars: immutable_1.Map(),
            foundationEventListeners: immutable_1.Map(),
            open: false,
        };
        _this.adapter = new adapter_1.FoundationAdapter();
        _this.foundation = new mdc_drawer_1.MDCTemporaryDrawerFoundation(_this.adapter.toObject());
        return _this;
    }
    Meta.prototype.getChildContext = function () {
        return {
            adapter: this.adapter,
        };
    };
    // Sync props and internal state
    Meta.prototype.componentWillReceiveProps = function (props) {
        if (props.open !== this.state.open) {
            if (props.open) {
                this.foundation.open();
            }
            else {
                this.foundation.close();
            }
        }
    };
    // Foundation lifecycle
    Meta.prototype.componentDidMount = function () {
        this.adapter.setContainerAdapter(new ContainerAdapterImpl(this));
        this.foundation.init();
    };
    Meta.prototype.componentWillUnmount = function () {
        this.foundation.destroy();
        this.adapter.setContainerAdapter(new adapter_1.ContainerAdapter());
    };
    Meta.prototype.getClassName = function (_props, state) {
        return classNames(exports.CLASS_NAME, state.foundationClasses.toJS());
    };
    Meta.prototype.renderNativeDOMProps = function () {
        return {
            cssVariables: this.state.foundationCssVars.toJS(),
            eventListeners: this.state.foundationEventListeners.toJS(),
        };
    };
    Meta.prototype.renderBaseClassName = function () {
        return exports.CLASS_NAME;
    };
    Meta.prototype.renderClassValues = function () {
        return [this.getClassName(this.props, this.state)];
    };
    Meta.displayName = "Container";
    Meta.childContextTypes = {
        adapter: PropTypes.instanceOf(adapter_1.FoundationAdapter),
    };
    Meta.defaultProps = {
        rtl: false,
        style: {},
    };
    return Meta;
}(base_1.ClassNameMetaBase));
exports.Meta = Meta;
var ContainerAdapterImpl = (function (_super) {
    __extends(ContainerAdapterImpl, _super);
    function ContainerAdapterImpl(element) {
        var _this = _super.call(this) || this;
        _this.element = element;
        return _this;
    }
    ContainerAdapterImpl.prototype.addClass = function (className) {
        this.element.setState(function (state) { return ({
            foundationClasses: state.foundationClasses.add(className),
        }); });
        // MDCTemporaryDrawerFoundation does not provide opening/closing event.
        // But we can assume open/close by adding/removing OPEN_CLASS_NAME
        if (className === OPEN_CLASS_NAME) {
            this.element.setState({
                open: true,
            });
            if (this.element.props.onOpenDrawer) {
                this.element.props.onOpenDrawer(this.element);
            }
        }
    };
    ContainerAdapterImpl.prototype.removeClass = function (className) {
        this.element.setState(function (state) { return ({
            foundationClasses: state.foundationClasses.remove(className),
        }); });
        // MDCTemporaryDrawerFoundation does not provide opening/closing event.
        // But we can assume open/close by adding/removing OPEN_CLASS_NAME
        if (className === OPEN_CLASS_NAME) {
            this.element.setState({
                open: false,
            });
            if (this.element.props.onCloseDrawer) {
                this.element.props.onCloseDrawer(this.element);
            }
        }
    };
    ContainerAdapterImpl.prototype.hasClass = function (className) {
        return util_1.includes(this.element.getClassName(this.element.props, this.element.state).split(/\s+/), className);
    };
    ContainerAdapterImpl.prototype.registerInteractionHandler = function (evt, handler) {
        this.element.setState(function (state) { return ({
            foundationEventListeners: state.foundationEventListeners.update(drawerUtil.remapEvent(evt, window), immutable_1.OrderedSet(), function (x) { return x.add(handler); }),
        }); });
    };
    ContainerAdapterImpl.prototype.deregisterInteractionHandler = function (evt, handler) {
        this.element.setState(function (state) { return ({
            foundationEventListeners: state.foundationEventListeners.update(drawerUtil.remapEvent(evt, window), immutable_1.OrderedSet(), function (x) { return x.delete(handler); }),
        }); });
    };
    ContainerAdapterImpl.prototype.registerTransitionEndHandler = function (handler) {
        var evt = "transitionend";
        this.registerInteractionHandler(evt, handler);
    };
    ContainerAdapterImpl.prototype.deregisterTransitionEndHandler = function (handler) {
        var evt = "transitionend";
        this.deregisterInteractionHandler(evt, handler);
    };
    ContainerAdapterImpl.prototype.registerDocumentKeydownHandler = function (handler) {
        document.addEventListener("keydown", handler);
    };
    ContainerAdapterImpl.prototype.deregisterDocumentKeydownHandler = function (handler) {
        document.removeEventListener("keydown", handler);
    };
    ContainerAdapterImpl.prototype.updateCssVariable = function (value) {
        this.element.setState(function (state) { return ({
            foundationCssVars: state.foundationCssVars.set(OPACITY_VAR_NAME, value),
        }); });
    };
    ContainerAdapterImpl.prototype.saveElementTabState = function (el) {
        drawerUtil.saveElementTabState(el);
    };
    ContainerAdapterImpl.prototype.restoreElementTabState = function (el) {
        drawerUtil.restoreElementTabState(el);
    };
    ContainerAdapterImpl.prototype.makeElementUntabbable = function (el) {
        el.setAttribute("tabindex", "-1");
    };
    ContainerAdapterImpl.prototype.isRtl = function () {
        return this.element.props.rtl || false;
    };
    return ContainerAdapterImpl;
}(adapter_1.ContainerAdapter));
var Container = (function (_super) {
    __extends(Container, _super);
    function Container() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Container.prototype.getMetaComponent = function () {
        return Meta;
    };
    Container.prototype.getMetaPropNames = function () {
        return [
            "className",
            "open",
            "rtl",
            "style",
            "onOpenDrawer",
            "onCloseDrawer",
        ];
    };
    Container.prototype.getChildComponent = function () {
        return "aside";
    };
    Container.Meta = Meta;
    return Container;
}(base_1.DefaultComponentBase));
exports.default = Container;
