// @react-mdc/radio/lib/NativeControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ReactDOM = require("react-dom");
var PropTypes = require("prop-types");
var base_1 = require("@react-mdc/base");
var util_1 = require("@react-mdc/base/lib/util");
var adapter_1 = require("./adapter");
var constants_1 = require("./constants");
exports.CLASS_NAME = constants_1.BASE_CLASS_NAME + "__native-control";
exports.propertyClassNames = {
    PREFIX: exports.CLASS_NAME,
};
/**
 * Radio input component
 */
var Meta = (function (_super) {
    __extends(Meta, _super);
    function Meta() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleChange = function (evt) {
            _this.defaultOnChange(evt);
        };
        return _this;
    }
    Meta.prototype.componentDidMount = function () {
        this.context.adapter.setNativeControlAdapter(new NativeControlAdapterImpl(this));
    };
    Meta.prototype.componentWillUnmount = function () {
        this.context.adapter.setNativeControlAdapter(new adapter_1.NativeControlAdapter());
    };
    Meta.prototype.renderBaseClassName = function () {
        return exports.CLASS_NAME;
    };
    Meta.prototype.renderProps = function (childProps) {
        var onChange = this.props.onChange;
        return __assign({}, _super.prototype.renderProps.call(this, childProps), { checked: this.context.adapter.isChecked() || undefined, onChange: util_1.eventHandlerDecorator(this.handleChange)(onChange || null) });
    };
    Meta.contextTypes = {
        adapter: PropTypes.instanceOf(adapter_1.FoundationAdapter).isRequired,
    };
    return Meta;
}(base_1.ClassNameMetaBase));
exports.Meta = Meta;
var NativeControlAdapterImpl = (function (_super) {
    __extends(NativeControlAdapterImpl, _super);
    function NativeControlAdapterImpl(element) {
        var _this = _super.call(this) || this;
        _this.element = element;
        return _this;
    }
    NativeControlAdapterImpl.prototype.getNativeControl = function () {
        return ReactDOM.findDOMNode(this.element);
    };
    NativeControlAdapterImpl.prototype.setDefaultOnChangeHandler = function (onChange) {
        this.element.defaultOnChange = onChange;
    };
    return NativeControlAdapterImpl;
}(adapter_1.NativeControlAdapter));
// Input with type="radio" as default
function RadioInput(props) {
    return (React.createElement("input", __assign({ type: "radio" }, props)));
}
var NativeControl = (function (_super) {
    __extends(NativeControl, _super);
    function NativeControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NativeControl.prototype.getMetaComponent = function () {
        return Meta;
    };
    NativeControl.prototype.getMetaPropNames = function () {
        return [
            "className",
            "onChange",
        ];
    };
    NativeControl.prototype.getChildComponent = function () {
        return RadioInput;
    };
    NativeControl.Meta = Meta;
    return NativeControl;
}(base_1.DefaultComponentBase));
exports.default = NativeControl;
