// @react-mdc/checkbox/lib/Container.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var mdc_animation_1 = require("@material/animation/dist/mdc.animation");
var mdc_checkbox_1 = require("@material/checkbox/dist/mdc.checkbox");
var immutable_1 = require("immutable");
var PropTypes = require("prop-types");
var base_1 = require("@react-mdc/base");
var adapter_1 = require("./adapter");
var constants_1 = require("./constants");
exports.CLASS_NAME = constants_1.BASE_CLASS_NAME;
/**
 * Checkbox input container component
 */
var Meta = (function (_super) {
    __extends(Meta, _super);
    function Meta(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            foundationClasses: immutable_1.OrderedSet(),
            foundationEventListeners: immutable_1.Map(),
        };
        // Event handler
        _this.handleChange = function (_evt) {
            if (_this.props.checked != null) {
                if (_this.foundation.isChecked() !== _this.props.checked) {
                    // Checked state should not be changed by foundation
                    _this.foundation.setChecked(_this.props.checked);
                }
            }
        };
        _this.adapter = new adapter_1.FoundationAdapter();
        _this.foundation = new mdc_checkbox_1.MDCCheckboxFoundation(_this.adapter.toObject());
        return _this;
    }
    Meta.prototype.getChildContext = function () {
        return {
            adapter: this.adapter,
        };
    };
    // Foundation lifecycle
    Meta.prototype.componentDidMount = function () {
        this.adapter.setContainerAdapter(new ContainerAdapterImpl(this));
        this.foundation.init();
        this.adapter.setDefaultOnChangeHandler(this.handleChange);
        if (this.props.checked != null) {
            this.foundation.setChecked(this.props.checked);
        }
        if (this.props.disabled != null) {
            this.foundation.setDisabled(this.props.disabled);
        }
    };
    Meta.prototype.componentWillUnmount = function () {
        this.foundation.destroy();
        this.adapter.setContainerAdapter(new adapter_1.ContainerAdapter());
    };
    // Sync props and internal state
    Meta.prototype.componentWillReceiveProps = function (props) {
        this.syncFoundation(props);
    };
    Meta.prototype.renderNativeDOMProps = function () {
        return {
            eventListeners: this.state.foundationEventListeners.toJS(),
        };
    };
    Meta.prototype.renderBaseClassName = function () {
        return exports.CLASS_NAME;
    };
    Meta.prototype.renderClassValues = function () {
        return [
            this.state.foundationClasses.toJS(),
        ];
    };
    Meta.prototype.syncFoundation = function (props) {
        if (props.checked != null && this.foundation.isChecked() !== this.props.checked) {
            this.foundation.setChecked(props.checked);
        }
        if (props.disabled != null && this.foundation.isDisabled() !== this.props.disabled) {
            this.foundation.setDisabled(props.disabled);
        }
        if (props.indeterminate != null && this.foundation.isIndeterminate() !== this.props.indeterminate) {
            this.foundation.setIndeterminate(props.indeterminate);
        }
    };
    Meta.childContextTypes = {
        adapter: PropTypes.instanceOf(adapter_1.FoundationAdapter),
    };
    return Meta;
}(base_1.ClassNameMetaBase));
exports.Meta = Meta;
var ContainerAdapterImpl = (function (_super) {
    __extends(ContainerAdapterImpl, _super);
    function ContainerAdapterImpl(element) {
        var _this = _super.call(this) || this;
        _this.element = element;
        return _this;
    }
    ContainerAdapterImpl.prototype.addClass = function (className) {
        this.element.setState(function (state) { return ({
            foundationClasses: state.foundationClasses.add(className),
        }); });
    };
    ContainerAdapterImpl.prototype.removeClass = function (className) {
        this.element.setState(function (state) { return ({
            foundationClasses: state.foundationClasses.remove(className),
        }); });
    };
    ContainerAdapterImpl.prototype.registerAnimationEndHandler = function (handler) {
        this.element.setState(function (state) { return ({
            foundationEventListeners: state.foundationEventListeners.update(mdc_animation_1.getCorrectEventName(window, "animationend"), immutable_1.OrderedSet(), function (x) { return x.add(handler); }),
        }); });
    };
    ContainerAdapterImpl.prototype.deregisterAnimationEndHandler = function (handler) {
        var evt = mdc_animation_1.getCorrectEventName(window, "animationend");
        if (this.element.state.foundationEventListeners.get(evt).includes(handler)) {
            this.element.setState(function (state) { return ({
                foundationEventListeners: state.foundationEventListeners.delete(evt),
            }); });
        }
    };
    ContainerAdapterImpl.prototype.forceLayout = function () {
        /* no-op */
    };
    ContainerAdapterImpl.prototype.isAttachedToDOM = function () {
        // Always true. Because we initialize foundation on componentDidMount
        return true;
    };
    ContainerAdapterImpl.prototype.isChecked = function () {
        return this.element.props.checked || null;
    };
    return ContainerAdapterImpl;
}(adapter_1.ContainerAdapter));
var Container = (function (_super) {
    __extends(Container, _super);
    function Container() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Container.prototype.getMetaComponent = function () {
        return Meta;
    };
    Container.prototype.getMetaPropNames = function () {
        return [
            "className",
            "checked",
            "disabled",
            "indeterminate",
        ];
    };
    Container.prototype.getChildComponent = function () {
        return "div";
    };
    Container.Meta = Meta;
    return Container;
}(base_1.DefaultComponentBase));
exports.default = Container;
