// preact-material-components/Card/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Button = _interopRequireDefault(require("../Button"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _MaterialComponent = _interopRequireDefault(require("../MaterialComponent"));

var _preact = require("preact");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

class Card extends _MaterialComponent.default {
  constructor() {
    super();
    this.componentName = 'card';
    this._mdcProps = ['stroked'];
  }

}

class CardActions extends _MaterialComponent.default {
  constructor() {
    super();
    this.componentName = 'card__actions';
    this._mdcProps = ['full-bleed'];
  }

}

class CardMedia extends _MaterialComponent.default {
  constructor() {
    super();
    this.componentName = 'card__media';
    this._mdcProps = ['square', '16-9'];
  }

}

class CardActionButton extends _Button.default {
  constructor() {
    super();
    this.componentName = 'card__action';
  }

  materialDom(props) {
    return (0, _preact.h)("button", _extends({
      className: "mdc-button mdc-card__action--button"
    }, props, {
      ref: this.setControlRef
    }), props.children);
  }

}

class CardActionIcons extends _MaterialComponent.default {
  constructor() {
    super();
    this.componentName = 'card__action-icons';
  }

}

class CardActionIcon extends _Icon.default {
  constructor() {
    super();
    this.componentName = 'card__action';
  }

  materialDom(props) {
    props.className = 'mdc-card__action--icon';
    return super.materialDom(props);
  }

}

class CardMediaContent extends _MaterialComponent.default {
  constructor() {
    super();
    this.componentName = 'card__media-content';
  }

}

Card.Actions = CardActions;
Card.ActionButton = CardActionButton;
Card.ActionIcons = CardActionIcons;
Card.ActionIcon = CardActionIcon;
Card.Media = CardMedia;
Card.CardMediaContent = CardMediaContent;
var _default = Card;
exports.default = _default;
