// @react-mdc/switch/lib/NativeControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var base_1 = require("@react-mdc/base");
var constants_1 = require("./constants");
exports.CLASS_NAME = constants_1.BASE_CLASS_NAME + "__native-control";
/**
 * Native control component
 */
var Meta = (function (_super) {
    __extends(Meta, _super);
    function Meta() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Meta.prototype.renderBaseClassName = function () {
        return exports.CLASS_NAME;
    };
    return Meta;
}(base_1.ClassNameMetaBase));
exports.Meta = Meta;
// Input with type="checkbox" as default
function CheckboxInput(props) {
    return (React.createElement("input", __assign({ type: "checkbox" }, props)));
}
var NativeControl = (function (_super) {
    __extends(NativeControl, _super);
    function NativeControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NativeControl.prototype.getMetaComponent = function () {
        return Meta;
    };
    NativeControl.prototype.getMetaPropNames = function () {
        return [
            "className",
        ];
    };
    NativeControl.prototype.getChildComponent = function () {
        return CheckboxInput;
    };
    NativeControl.Meta = Meta;
    return NativeControl;
}(base_1.DefaultComponentBase));
exports.default = NativeControl;
